### PR TITLE
python3Packages.black: 20.8b1 -> 21.4b2

### PR DESCRIPTION
**Pull request by sveitser**
###### Motivation for this change

Package update. 

There are 2 new persistently failing tests (currently disabled) where I don't know why they fail.

```
=================================== FAILURES ===================================
_________ BlackTestCase.test_output_locking_when_writeback_color_diff __________

self = <tests.test_black.BlackTestCase testMethod=test_output_locking_when_writeback_color_diff>

    @event_loop()
    def test_output_locking_when_writeback_color_diff(self) -> None:
        with cache_dir() as workspace:
            for tag in range(0, 4):
                src = (workspace / f"test{tag}.py").resolve()
                with src.open("w") as fobj:
                    fobj.write("print('hello')")
            with patch("black.Manager", wraps=multiprocessing.Manager) as mgr:
                self.invokeBlack(["--diff", "--color", str(workspace)], exit_code=0)
                # this isn't quite doing what we want, but if it _isn't_
                # called then we cannot be using the lock it provides
>               mgr.assert_called()

tests/test_black.py:1125:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <MagicMock name='Manager' id='140737305834208'>

    def assert_called(self):
        """assert that the mock was called at least once
        """
        if self.call_count == 0:
            msg = ("Expected '%s' to have been called." %
                   (self._mock_name or 'mock'))
>           raise AssertionError(msg)
E           AssertionError: Expected 'Manager' to have been called.

/nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/python3.8/unittest/mock.py:882: AssertionError
____________ BlackTestCase.test_output_locking_when_writeback_diff _____________

self = <tests.test_black.BlackTestCase testMethod=test_output_locking_when_writeback_diff>

    @event_loop()
    def test_output_locking_when_writeback_diff(self) -> None:
        with cache_dir() as workspace:
            for tag in range(0, 4):
                src = (workspace / f"test{tag}.py").resolve()
                with src.open("w") as fobj:
                    fobj.write("print('hello')")
            with patch("black.Manager", wraps=multiprocessing.Manager) as mgr:
                self.invokeBlack(["--diff", str(workspace)], exit_code=0)
                # this isn't quite doing what we want, but if it _isn't_
                # called then we cannot be using the lock it provides
>               mgr.assert_called()

tests/test_black.py:1112:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <MagicMock name='Manager' id='140737305116240'>

    def assert_called(self):
        """assert that the mock was called at least once
        """
        if self.call_count == 0:
            msg = ("Expected '%s' to have been called." %
                   (self._mock_name or 'mock'))
>           raise AssertionError(msg)
E           AssertionError: Expected 'Manager' to have been called.

/nix/store/q6gfck5czr67090pwm53xrdyhpg6bx67-python3-3.8.9/lib/python3.8/unittest/mock.py:882: AssertionError
=========================== short test summary info ============================
FAILED tests/test_black.py::BlackTestCase::test_output_locking_when_writeback_color_diff
FAILED tests/test_black.py::BlackTestCase::test_output_locking_when_writeback_diff
=========== 2 failed, 174 passed, 4 deselected, 3 xfailed in 34.59s ============
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sveitser** [comment]
Result of `nixpkgs-review pr 120684` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages marked as broken and skipped:</summary>
  <ul>
    <li>bareos</li>
    <li>pepper</li>
    <li>python39Packages.hypothesmith</li>
    <li>samba4Full</li>
    <li>tribler</li>
  </ul>
</details>
<details>
  <summary>7 packages failed to build:</summary>
  <ul>
    <li>python38Packages.pyls-black</li>
    <li>python38Packages.pylxd</li>
    <li>spyder (python38Packages.spyder)</li>
    <li>python38Packages.stumpy</li>
    <li>python39Packages.pyls-black</li>
    <li>python39Packages.pylxd</li>
    <li>python39Packages.spyder</li>
  </ul>
</details>
<details>
  <summary>61 packages built:</summary>
  <ul>
    <li>black (python38Packages.black)</li>
    <li>black-macchiato (python38Packages.black-macchiato)</li>
    <li>ceph</li>
    <li>ceph-client</li>
    <li>errbot</li>
    <li>fava</li>
    <li>flexget</li>
    <li>libceph</li>
    <li>mnemosyne</li>
    <li>noto-fonts-emoji</li>
    <li>python38Packages.WSME</li>
    <li>python38Packages.backports_functools_lru_cache</li>
    <li>python38Packages.cheroot</li>
    <li>python38Packages.cherrypy</li>
    <li>python38Packages.diagrams</li>
    <li>python38Packages.graphql-server-core</li>
    <li>python38Packages.hypothesmith</li>
    <li>python38Packages.irc</li>
    <li>python38Packages.jaraco_collections</li>
    <li>python38Packages.jaraco_functools</li>
    <li>python38Packages.jaraco_logging</li>
    <li>python38Packages.jaraco_text</li>
    <li>python38Packages.nototools</li>
    <li>python38Packages.papermill</li>
    <li>python38Packages.portend</li>
    <li>python38Packages.ptpython</li>
    <li>python38Packages.pyalgotrade</li>
    <li>python38Packages.pytest-black</li>
    <li>python38Packages.scancode-toolkit</li>
    <li>sqlite-utils (python38Packages.sqlite-utils)</li>
    <li>python38Packages.tempora</li>
    <li>python38Packages.typer</li>
    <li>python38Packages.web</li>
    <li>python38Packages.ws4py</li>
    <li>python39Packages.backports_functools_lru_cache</li>
    <li>python39Packages.black</li>
    <li>python39Packages.black-macchiato</li>
    <li>python39Packages.cheroot</li>
    <li>python39Packages.cherrypy</li>
    <li>python39Packages.diagrams</li>
    <li>python39Packages.graphql-server-core</li>
    <li>python39Packages.irc</li>
    <li>python39Packages.jaraco_collections</li>
    <li>python39Packages.jaraco_functools</li>
    <li>python39Packages.jaraco_logging</li>
    <li>python39Packages.jaraco_text</li>
    <li>python39Packages.nototools</li>
    <li>python39Packages.papermill</li>
    <li>python39Packages.portend</li>
    <li>python39Packages.ptpython</li>
    <li>python39Packages.pyalgotrade</li>
    <li>python39Packages.pytest-black</li>
    <li>python39Packages.scancode-toolkit</li>
    <li>python39Packages.sqlite-utils</li>
    <li>python39Packages.tempora</li>
    <li>python39Packages.typer</li>
    <li>python39Packages.web</li>
    <li>python39Packages.ws4py</li>
    <li>qemu_full</li>
    <li>sabnzbd</li>
    <li>twitter-color-emoji</li>
  </ul>
</details>

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 120684` at b068b351 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages marked as broken and skipped:</summary>
  <ul>
    <li>bareos</li>
    <li>pepper</li>
    <li>python38Packages.stumpy</li>
    <li>python39Packages.hypothesmith</li>
    <li>tribler</li>
  </ul>
</details>
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/33f784bc4b3acbb4fba6b91519c41200">python38Packages.pyls-black</a></li>
    <li><a href="https://gist.github.com/eef60a5a0fbbd863bab451105619d54d">python38Packages.pylxd</a></li>
    <li><a href="https://gist.github.com/5bf5332308b7b6fc71563b2248a62ef6">spyder</a> (python38Packages.spyder)</li>
    <li><a href="https://gist.github.com/9f2237d06036ac8c515364b5a734f64f">python39Packages.pyls-black</a></li>
    <li><a href="https://gist.github.com/ee4600c325d0fd484665d8189327ed09">python39Packages.pylxd</a></li>
    <li><a href="https://gist.github.com/17e7411b94bad6466a0181c3f3e941bc">python39Packages.spyder</a></li>
  </ul>
</details>
<details>
  <summary>5 packages skipped due to time constraints:</summary>
  <ul>
    <li>ceph (ceph-dev)</li>
    <li>ceph-client</li>
    <li>libceph</li>
    <li>noto-fonts-emoji</li>
    <li>qemu_full</li>
  </ul>
</details>
<details>
  <summary>55 packages built successfully:</summary>
  <ul>
    <li>black (python38Packages.black)</li>
    <li>black-macchiato (python38Packages.black-macchiato)</li>
    <li>errbot</li>
    <li>fava</li>
    <li>flexget</li>
    <li>mnemosyne</li>
    <li>python38Packages.WSME</li>
    <li>python38Packages.backports_functools_lru_cache</li>
    <li>python38Packages.cheroot</li>
    <li>python38Packages.cherrypy</li>
    <li>python38Packages.diagrams</li>
    <li>python38Packages.graphql-server-core</li>
    <li>python38Packages.hypothesmith</li>
    <li>python38Packages.irc</li>
    <li>python38Packages.jaraco_collections</li>
    <li>python38Packages.jaraco_functools</li>
    <li>python38Packages.jaraco_logging</li>
    <li>python38Packages.jaraco_text</li>
    <li>python38Packages.nototools</li>
    <li>python38Packages.papermill</li>
    <li>python38Packages.portend</li>
    <li>python38Packages.ptpython</li>
    <li>python38Packages.pyalgotrade</li>
    <li>python38Packages.pytest-black</li>
    <li>python38Packages.scancode-toolkit</li>
    <li>sqlite-utils (python38Packages.sqlite-utils)</li>
    <li>python38Packages.tempora</li>
    <li>python38Packages.typer</li>
    <li>python38Packages.web</li>
    <li>python38Packages.ws4py</li>
    <li>python39Packages.backports_functools_lru_cache</li>
    <li>python39Packages.black</li>
    <li>python39Packages.black-macchiato</li>
    <li>python39Packages.cheroot</li>
    <li>python39Packages.cherrypy</li>
    <li>python39Packages.diagrams</li>
    <li>python39Packages.graphql-server-core</li>
    <li>python39Packages.irc</li>
    <li>python39Packages.jaraco_collections</li>
    <li>python39Packages.jaraco_functools</li>
    <li>python39Packages.jaraco_logging</li>
    <li>python39Packages.jaraco_text</li>
    <li>python39Packages.nototools</li>
    <li>python39Packages.papermill</li>
    <li>python39Packages.portend</li>
    <li>python39Packages.ptpython</li>
    <li>python39Packages.pyalgotrade</li>
    <li>python39Packages.pytest-black</li>
    <li>python39Packages.scancode-toolkit</li>
    <li>python39Packages.sqlite-utils</li>
    <li>python39Packages.tempora</li>
    <li>python39Packages.typer</li>
    <li>python39Packages.web</li>
    <li>python39Packages.ws4py</li>
    <li>sabnzbd</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement.

**Comment by sveitser** [comment]
- `pylxd` also failis with `ERROR: No matching distribution found for cryptography>=3.2` on master.
- `spyder` depends on `pyls-black`.

**Comment by dotlambda** [comment]
> pylxd also fails with ERROR: No matching distribution found for cryptography>=3.2 on master.

fixed in https://github.com/NixOS/nixpkgs/pull/120704

**Comment by SuperSandro2000** [comment]
Could it be that they are writting into a write only directory?

**Comment by sveitser** [comment]
> Could it be that they are writting into a write only directory?

Yes. I think the problem is that the `get_sources` function here https://github.com/psf/black/blob/04fd4432f6c6007e419d7174930569a75ea60fed/src/black/__init__.py#L618 doesn't return any paths for these two test cases. As a result black just exits with ("nothing to do") and the lock method of the manager is also never called.

**Comment by sveitser** [comment]
The files are being ignored because we are running in the `/build/` directory and that matches the `DEFAULT_EXCLUDES` regex.

```python
DEFAULT_EXCLUDES = r"/(\.direnv|\.eggs|\.git|\.hg|\.mypy_cache|\.nox|\.tox|\.venv|venv|\.svn|_build|buck-out|build|dist)/"  # noqa: B950
```

https://github.com/psf/black/blob/04fd4432f6c6007e419d7174930569a75ea60fed/src/black/__init__.py#L85

**Comment by SuperSandro2000** [comment]
> The files are being ignored because we are running in the `/build/` directory and that matches the `DEFAULT_EXCLUDES` regex.

Can we fix that? Or is that to much work?

**Comment by sveitser** [comment]
We could for example add a `/build/.git` file to make black detect `/build` as the project root.

```nix
preCheck = ''
    export PATH="$PATH:$out/bin"

    # Make /build the project root for black tests.
    # If the paths contain /build they match black's DEFAULT_EXCLUDE regex.
    touch /build/.git
  '';
```
Not sure if this is a good idea. I noticed this also makes 3 of the 4 currently failing tests pass.

The project root is detected with this function.

https://github.com/psf/black/blob/04fd4432f6c6007e419d7174930569a75ea60fed/src/black/__init__.py#L6272-L6308

**Comment by SuperSandro2000** [comment]
> Not sure if this is a good idea. I noticed this also makes 3 of the 4 currently failing tests pass.

If you don't know a nicer way to fix that we can do it like this. It's always better to be a little hacky then to not run any tests at all.

File: pkgs/development/python-modules/black/default.nix [hunk 1]
```diff
@@ -30,19 +31,19 @@ buildPythonPackage rec {
...
 
-  checkInputs =  [ pytestCheckHook ];
+  checkInputs =  [ pytestCheckHook parameterized ];
 
   preCheck = ''
     export PATH="$PATH:$out/bin"
+
+    # The top directory /build matches black's DEFAULT_EXCLUDE regex.
+    # Make /build the project root for black tests to avoid excluding files.
+    touch /build/.git
```
Comment (dotlambda) [inline pkgs/development/python-modules/black/default.nix:41]:
```suggestion
    touch ../.git
```

**Review by dotlambda — CHANGES_REQUESTED** [review]
Please squash the commits into one.

File: pkgs/development/python-modules/black/default.nix (hunk 1 above)
Comment (sveitser) [reply pkgs/development/python-modules/black/default.nix]: Applied. Thanks.

**Comment by dotlambda** [comment]
Can you include your fix from https://github.com/rupert/pyls-black/pull/39 using `fetchpatch` in a separate commit?

File: pkgs/development/python-modules/pyls-black/default.nix [hunk 2]
```diff
@@ -13,6 +13,15 @@ buildPythonPackage rec {
     sha256 = "0cjf0mjn156qp0x6md6mncs31hdpzfim769c2lixaczhyzwywqnj";
   };
 
+  # Fix test failure with black 21.4b0+
+  # Remove if https://github.com/rupert/pyls-black/pull/39 merged.
+  patches = [
+    (fetchpatch {
+      url = "https://github.com/rupert/pyls-black/commit/728207b540d9c25eb0d1cd96419ebfda2e257f63.patch";
+      sha256 = "044zpkwjfjnslhl2jq0fn74fw2r556gmwsjdq88mbb0jph9lm3xs";
```
Comment (dotlambda) [inline pkgs/development/python-modules/pyls-black/default.nix:21]:
```suggestion
      sha256 = "0i3w5myhjl5lq1lpkizagnmk6m8fkn3igfyv5f2qcrn5n7f119ak";
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Comment by dotlambda** [comment]
Please squash the last two commits.

**Pull request merged**